java/src/main/java/ai/rapids/cudf/MemoryCleaner.java

```diff
@@ -199,6 +200,12 @@ static void setDefaultGpu(int defaultGpuId) {
             log.error("CAUGHT EXCEPTION WHILE TRYING TO CLEAN " + next, t);
           }
           all.remove(next);
+        } else {
+          long currentTime = System.currentTimeMillis();
+          if (currentTime - lastRemoveCleanedObjectsTime > 10 * 1000L) { // every 10 seconds
+            all.removeIf(e -> e.cleaner.isClean());
```

Comment:
Why are we polling to clean objects when we can proactively remove things from `all` when they are properly closed?  Saves from doing a full scan of potentially very many live objects in `all` and should be more efficient.

There's a couple of ways to do this.  One is to implement equals and hashcode for `CleanerWeakReference` so we can remove our cleaner from all (by building a temporary instance to do the remove).  More efficiently would be to rework the `all` and Cleaner setups so instead of `CleanerWeakReference` pointing to `Cleaner`, we remove `CleanerWeakReference` entirely and have `Cleaner` have the `isRmmBlocker` field a normal `WeakReference` to the object it is associated with.  `all` would simply be a hash set of the cleaner instances, and then it's trivial for a cleaner to remove itself from `all` when it cleans.

Reply:
Done.
Now proactively remove.

Reply:
Use cleaner id as the key of the Map. The id is unique and can represent a unique resource.